### PR TITLE
Add build dependency on fmtlib

### PR DESCRIPTION
**Pull request by joakim-hove**
In C++20 the library `fmt`which replaces `printf()` and `<<` based formatting - looks very good. 

Before C++20 we can use the external library: https://github.com/fmtlib/fmt/ - which is essentially what was included for C++20. 

For the ongoing work with improvements in error messages this would be extremely valuable, I therefor suggest that we use `fmtlib` until we switch to C++20. 

This PR assumes that fmtlib is installed on your system (it is available as binary package on RH, Ubuntu & Debian) and integrates that in the build system - the cmake integration is quite naive and probably needs some love and attention from someone in the know.

### CODE REVIEW

File: opm-common-prereqs.cmake [hunk 1]
```diff
@@ -15,6 +15,7 @@ set (opm-common_DEPS
 list(APPEND opm-common_DEPS
       # various runtime library enhancements
       "Boost 1.44.0 COMPONENTS system unit_test_framework REQUIRED"
+			"fmt"
```
Comment (akva2) [inline opm-common-prereqs.cmake:18]: fmt is REQUIRED and watch the tabs.

Comment (joakim-hove) [reply opm-common-prereqs.cmake]: Fixed

**Comment by alfbr** [comment]
This is potentially very painful for us. Can you please elaborate how this is available in Red Hat (e.g., does it suffice to use a devtoolset when compilinng for Red Hat 6 and 7?).

**Comment by akva2** [comment]
it's in epel.

**Comment by alfbr** [comment]
> it's in epel.

If it is installed from EPEL and linked dynamically, this quickly means we need it installed on everyting, not just build servers.

**Comment by akva2** [comment]
fmt is a pure template library.

**Comment by alfbr** [comment]
> fmt is a pure template library.

Only a compile time dependency then? That makes it much less painful. It seems like it is available for both Red Hat 6 and 7 in EPEL. Sounds like it will be part of the GCC toolchain come C++20. If my understanding is correct, this is fine from my end.

**Comment by akva2** [comment]
precisely.

**Comment by atgeirr** [comment]
If it is a header-only lib, why do we change the target_link_libraries() call?

**Comment by akva2** [comment]
because it pulls in include paths from the fmt::fmt dependency.

**Comment by atgeirr** [comment]
I assumed target_link_libraries() only added LINK information. What you say seems to imply that it adds COMPILE information as well, is that correct?

**Comment by akva2** [comment]
yes, fmt uses cmake targets, which is the "modern" way of doing cmake. these are cool because they carry all info. including include paths. the target_link_libraries call adds the include path as well. no need for separate _LIBRARES, _INCLUDE_DIRS etc variables any more.

**Comment by bska** [comment]
> fmt uses cmake targets, which is the "modern" way of doing cmake

True.  If we ever decide to take the time to rework OPM's CMake-based build system, then targets and properties is the way to do it.

**Comment by atgeirr** [comment]
Thanks for the information. Now I am not against merging this, but I need one more piece of information: What do I need to do on a Ubuntu system to get this? I ask because we must update the build instructions on the website. (I assume `brew install <something>` is good enough on my mac.)

**Comment by akva2** [comment]
```apt install libfmt-dev```

File: CMakeLists.txt [hunk 2]
```diff
@@ -373,3 +373,5 @@ if (OPM_ENABLE_PYTHON)
     add_definitions(-DEMBEDDED_PYTHON)
   endif()
 endif()
+
+target_link_libraries(opmcommon PUBLIC fmt::fmt)
```
Comment (bska) [inline CMakeLists.txt:377]: Looking at the [documentation](https://fmt.dev/latest/usage.html#building-the-library) this line could **arguably** be
```cmake
target_link_library(opmcommon PRIVATE fmt::fmt-header-only)
```
instead.  The above appears to link to a precompiled (`.a` or `.so`) version of the `fmt` library.

Comment (joakim-hove) [reply CMakeLists.txt:377]: Tried that - did not get it to work.

Comment (bska) [reply CMakeLists.txt:377]: > Tried that - did not get it to work.

In that case this (using `fmt`) may be more untenable.  If we do indeed need a precompiled third-party library then it is a much more expensive proposition.

Comment (joakim-hove) [reply CMakeLists.txt:377]: Well - that debian package (which I am using) has `libfmt.a` file which is linked to - but that is a compile time dependency - so I do not see a problem with that?

Comment (akva2) [reply CMakeLists.txt:377]: it's a shared object on epel. anyways, see https://github.com/akva2/opm-common/commit/d8f0106ee3d5169215107fd28987b6bf146c6f6d (sorry you have to "distribute" across your commits).

tested on rhel and ubuntu.

the fmt-header-only target is fairly recent so we cannot rely on it (even ubuntu 20.04 is on fmt 6.1.2 which does not export it in the config files).

Comment (joakim-hove) [reply CMakeLists.txt:377]: Thank'll pick that up; but it does look like you have chosen a non optimal starting point for your fixes. I guess if you rebase on top of the tip of this branch I can just merge the changes right in?

Comment (akva2) [reply CMakeLists.txt:377]: no, it sits on top of two of your branches and changes are across different commits. just make it happen, i don't care about authorship or whatever.

Comment (joakim-hove) [reply CMakeLists.txt]: OK - done; I have added `add_definitions(-DFMT_HEADER_ONLY)`. Thank you

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
OK; I see this as roughly ack'ed - I will merge now. All developers need to install fmtlib on their machine. Unfortunately there are some version differences between the fmtlib versions available for different distributions, hopefully this will not create problems.

I will merge this now, but will wait a bit with dependent code - so that we can easily pull it out again if it creates problems.

**Comment by atgeirr** [comment]
I am struggling, but will try a bit more before I ask for mercy. I think a notice to the mailing list would be appropriate?

**Comment by joakim-hove** [comment]
> I think a notice to the mailing list would be appropriate?

I agree

**Comment by atgeirr** [comment]
I managed to figure out things. There were no installation instructions, but a normal cmake-based build process was sufficient.

**Pull request merged**